File: apps/desktop/src-tauri/src/commands/bureau.rs
use crate::db::{
    Competition, CompetitionEntry, CompetitionTeam, CreateCompetition, CreatePerson,
    EntryResultDetail, EntryResultSummary, Person, SeriesResultSummary, TeamResultSummary,
};
use crate::engine::StandEngine;
use std::sync::Arc;

#[tauri::command]
pub fn list_people(
    engine: tauri::State<'_, Arc<StandEngine>>,
    query: Option<String>,
    include_archived: Option<bool>,
) -> Result<Vec<Person>, String> {
    engine.with_db(|db| db.list_people(query.as_deref(), include_archived.unwrap_or(false)))
}

#[tauri::command]
pub fn create_person(
    engine: tauri::State<'_, Arc<StandEngine>>,
    person: CreatePerson,
) -> Result<Person, String> {
    engine.with_db(|db| db.create_person(person))
}

#[tauri::command]
pub fn set_person_archived(
    engine: tauri::State<'_, Arc<StandEngine>>,
    id: String,
    archived: bool,
) -> Result<Person, String> {
    engine.with_db(|db| db.set_person_archived(&id, archived))
}

#[tauri::command]
pub fn update_person(
    engine: tauri::State<'_, Arc<StandEngine>>,
    id: String,
    person: CreatePerson,
) -> Result<Person, String> {
    engine.with_db(|db| db.update_person(&id, person))
}

#[tauri::command]
pub fn delete_person(
    engine: tauri::State<'_, Arc<StandEngine>>,
    id: String,
) -> Result<(), String> {
    engine.with_db(|db| db.delete_person(&id))
}

#[tauri::command]
pub fn list_competitions(
    engine: tauri::State<'_, Arc<StandEngine>>,
    include_archived: Option<bool>,
) -> Result<Vec<Competition>, String> {
    engine.with_db(|db| db.list_competitions(include_archived.unwrap_or(false)))
}

#[tauri::command]
pub fn create_competition(
    engine: tauri::State<'_, Arc<StandEngine>>,
    competition: CreateCompetition,
) -> Result<Competition, String> {
    engine.with_db(|db| db.create_competition(competition))
}

#[tauri::command]
pub fn update_competition(
    engine: tauri::State<'_, Arc<StandEngine>>,
    id: String,
    competition: CreateCompetition,
) -> Result<Competition, String> {
    engine.with_db(|db| db.update_competition(&id, competition))
}

#[tauri::command]
pub fn set_competition_status(
    engine: tauri::State<'_, Arc<StandEngine>>,
    id: String,
    status: String,
) -> Result<Competition, String> {
    engine.with_db(|db| db.set_competition_status(&id, &status))
}

#[tauri::command]
pub fn create_from_competition(
    engine: tauri::State<'_, Arc<StandEngine>>,
    source_id: String,
    name: Option<String>,
    date: Option<String>,
    as_template: Option<bool>,
    copy_entries: Option<bool>,
) -> Result<Competition, String> {
    engine.with_db(|db| {
        db.create_from_competition(
            &source_id,
            name.as_deref(),
            date.as_deref(),
            as_template.unwrap_or(false),
            copy_entries.unwrap_or(true),
        )
    })
}

#[tauri::command]
pub fn set_competition_team_settings(
    engine: tauri::State<'_, Arc<StandEngine>>,
    id: String,
    team_scoring_enabled: bool,
    team_count: i64,
) -> Result<Competition, String> {
    engine.with_db(|db| db.set_competition_team_settings(&id, team_scoring_enabled, team_count))
}

#[tauri::command]
pub fn list_entries(
    engine: tauri::State<'_, Arc<StandEngine>>,
    competition_id: String,
) -> Result<Vec<CompetitionEntry>, String> {
    engine.with_db(|db| db.list_entries(&competition_id))
}

#[tauri::command]
pub fn add_entry(
    engine: tauri::State<'_, Arc<StandEngine>>,
    competition_id: String,
    person_id: String,
) -> Result<CompetitionEntry, String> {
    engine.with_db(|db| db.add_entry(&competition_id, &person_id))
}

#[tauri::command]
pub fn reorder_entries(
    engine: tauri::State<'_, Arc<StandEngine>>,
    competition_id: String,
    entry_ids: Vec<String>,
) -> Result<Vec<CompetitionEntry>, String> {
    engine.with_db(|db| db.reorder_entries(&competition_id, &entry_ids))
}

#[tauri::command]
pub fn set_entry_status(
    engine: tauri::State<'_, Arc<StandEngine>>,
    entry_id: String,
    status: String,
) -> Result<CompetitionEntry, String> {
    engine.with_db(|db| db.set_entry_status(&entry_id, &status))
}

/// Deprecated no-op: Nachkauf series counter is incremented on start, not via this command.
#[tauri::command]
pub fn set_entry_nachkauf(
    engine: tauri::State<'_, Arc<StandEngine>>,
    entry_id: String,
    nachkauf_purchased: i64,
) -> Result<CompetitionEntry, String> {
    engine.with_db(|db| db.set_entry_nachkauf(&entry_id, nachkauf_purchased))
}

#[tauri::command]
pub fn remove_entry(
    engine: tauri::State<'_, Arc<StandEngine>>,
    entry_id: String,
) -> Result<(), String> {
    engine.with_db(|db| db.remove_entry(&entry_id))
}

#[tauri::command]
pub fn clone_entries(
    engine: tauri::State<'_, Arc<StandEngine>>,
    from_competition_id: String,
    to_competition_id: String,
) -> Result<Vec<CompetitionEntry>, String> {
    engine.with_db(|db| db.clone_entries_from(&from_competition_id, &to_competition_id))
}

#[tauri::command]
pub fn list_competition_results(
    engine: tauri::State<'_, Arc<StandEngine>>,
    competition_id: String,
) -> Result<Vec<EntryResultSummary>, String> {
    engine.with_db(|db| db.list_competition_results(&competition_id))
}

#[tauri::command]
pub fn get_entry_result(
    engine: tauri::State<'_, Arc<StandEngine>>,
    entry_id: String,
) -> Result<Option<EntryResultDetail>, String> {
    engine.with_db(|db| db.get_entry_result(&entry_id))
}

#[tauri::command]
pub fn list_entry_series(
    engine: tauri::State<'_, Arc<StandEngine>>,
    entry_id: String,
) -> Result<Vec<SeriesResultSummary>, String> {
    engine.with_db(|db| db.list_entry_series(&entry_id))
}

#[tauri::command]
pub fn list_teams(
    engine: tauri::State<'_, Arc<StandEngine>>,
    competition_id: Option<String>,
    include_archived: Option<bool>,
) -> Result<Vec<CompetitionTeam>, String> {
    engine.with_db(|db| {
        db.list_teams(
            competition_id.as_deref(),
            include_archived.unwrap_or(false),
        )
    })
}

#[tauri::command]
pub fn list_known_team_names(
    engine: tauri::State<'_, Arc<StandEngine>>,
    include_archived: Option<bool>,
) -> Result<Vec<String>, String> {
    engine.with_db(|db| db.list_known_team_names(include_archived.unwrap_or(false)))
}

#[tauri::command]
pub fn create_team(
    engine: tauri::State<'_, Arc<StandEngine>>,
    name: String,
    // Ignored — teams are global. Kept for older frontend payloads.
    #[allow(unused_variables)]
    competition_id: Option<String>,
) -> Result<CompetitionTeam, String> {
    engine.with_db(|db| db.create_team(&name))
}

#[tauri::command]
pub fn rename_team(
    engine: tauri::State<'_, Arc<StandEngine>>,
    team_id: String,
    name: String,
) -> Result<CompetitionTeam, String> {
    engine.with_db(|db| db.rename_team(&team_id, &name))
}

#[tauri::command]
pub fn set_team_archived(
    engine: tauri::State<'_, Arc<StandEngine>>,
    team_id: String,
    archived: bool,
) -> Result<CompetitionTeam, String> {
    engine.with_db(|db| db.set_team_archived(&team_id, archived))
}

#[tauri::command]
pub fn remove_team(
    engine: tauri::State<'_, Arc<StandEngine>>,
    team_id: String,
) -> Result<(), String> {
    engine.with_db(|db| db.remove_team(&team_id))
}

#[tauri::command]
pub fn add_team_member(
    engine: tauri::State<'_, Arc<StandEngine>>,
    team_id: String,
    entry_id: String,
) -> Result<CompetitionTeam, String> {
    engine.with_db(|db| db.add_team_member(&team_id, &entry_id))
}

#[tauri::command]
pub fn remove_team_member(
    engine: tauri::State<'_, Arc<StandEngine>>,
    team_id: String,
    entry_id: String,
) -> Result<CompetitionTeam, String> {
    engine.with_db(|db| db.remove_team_member(&team_id, &entry_id))
}

#[tauri::command]
pub fn add_team_person(
    engine: tauri::State<'_, Arc<StandEngine>>,
    team_id: String,
    person_id: String,
) -> Result<CompetitionTeam, String> {
    engine.with_db(|db| db.add_team_person(&team_id, &person_id))
}

#[tauri::command]
pub fn remove_team_person(
    engine: tauri::State<'_, Arc<StandEngine>>,
    team_id: String,
    person_id: String,
) -> Result<CompetitionTeam, String> {
    engine.with_db(|db| db.remove_team_person(&team_id, &person_id))
}

#[tauri::command]
pub fn list_team_results(
    engine: tauri::State<'_, Arc<StandEngine>>,
    competition_id: String,
) -> Result<Vec<TeamResultSummary>, String> {
    engine.with_db(|db| db.list_team_results(&competition_id))
}

File: apps/desktop/src/api/bureau.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  Competition,
  CompetitionEntry,
  CompetitionStatus,
  CompetitionTeam,
  EntryStatus,
  Person,
  ScoringMode,
  TeamResultSummary,
} from "@rotpunktarena/domain";

export async function listPeople(
  query?: string,
  includeArchived = false,
): Promise<Person[]> {
  return invoke("list_people", {
    query: query ?? null,
    includeArchived,
  });
}

export async function createPerson(input: {
  firstName: string;
  lastName: string;
  club?: string | null;
}): Promise<Person> {
  return invoke("create_person", { person: input });
}

export async function updatePerson(
  id: string,
  input: {
    firstName: string;
    lastName: string;
    club?: string | null;
  },
): Promise<Person> {
  return invoke("update_person", { id, person: input });
}

export async function deletePerson(id: string): Promise<void> {
  await invoke("delete_person", { id });
}

export async function setPersonArchived(
  id: string,
  archived: boolean,
): Promise<Person> {
  return invoke("set_person_archived", { id, archived });
}

export async function listCompetitions(
  includeArchived = false,
): Promise<Competition[]> {
  return invoke("list_competitions", { includeArchived });
}

export async function createCompetition(input: {
  name: string;
  date: string;
  discipline: string;
  maxShots: number;
  scoringMode: ScoringMode;
  nachkaufEnabled?: boolean;
  nachkaufShots?: number;
  teamScoringEnabled?: boolean;
  teamCount?: number;
  kind?: import("@rotpunktarena/domain").CompetitionKind;
  tenthsEnabled?: boolean;
  probeEnabled?: boolean;
}): Promise<Competition> {
  return invoke("create_competition", { competition: input });
}

export async function updateCompetition(
  id: string,
  input: {
    name: string;
    date: string;
    discipline: string;
    maxShots: number;
    scoringMode: ScoringMode;
    nachkaufEnabled?: boolean;
    nachkaufShots?: number;
    teamScoringEnabled?: boolean;
    teamCount?: number;
    kind?: import("@rotpunktarena/domain").CompetitionKind;
    tenthsEnabled?: boolean;
    probeEnabled?: boolean;
  },
): Promise<Competition> {
  return invoke("update_competition", { id, competition: input });
}

export async function createFromCompetition(input: {
  sourceId: string;
  name?: string | null;
  date?: string | null;
  asTemplate?: boolean;
  copyEntries?: boolean;
}): Promise<Competition> {
  return invoke("create_from_competition", {
    sourceId: input.sourceId,
    name: input.name ?? null,
    date: input.date ?? null,
    asTemplate: input.asTemplate ?? false,
    copyEntries: input.copyEntries ?? true,
  });
}

export async function setCompetitionStatus(
  id: string,
  status: CompetitionStatus,
): Promise<Competition> {
  return invoke("set_competition_status", { id, status });
}

export async function setCompetitionTeamSettings(
  id: string,
  teamScoringEnabled: boolean,
  teamCount: number,
): Promise<Competition> {
  return invoke("set_competition_team_settings", {
    id,
    teamScoringEnabled,
    teamCount,
  });
}

export async function listEntries(competitionId: string): Promise<CompetitionEntry[]> {
  return invoke("list_entries", { competitionId });
}

export async function addEntry(
  competitionId: string,
  personId: string,
): Promise<CompetitionEntry> {
  return invoke("add_entry", { competitionId, personId });
}

export async function reorderEntries(
  competitionId: string,
  entryIds: string[],
): Promise<CompetitionEntry[]> {
  return invoke("reorder_entries", { competitionId, entryIds });
}

export async function setEntryStatus(
  entryId: string,
  status: EntryStatus,
): Promise<CompetitionEntry> {
  return invoke("set_entry_status", { entryId, status });
}

/** @deprecated No-op; Nachkauf series counter increments on start. */
export async function setEntryNachkauf(
  entryId: string,
  nachkaufPurchased: number,
): Promise<CompetitionEntry> {
  return invoke("set_entry_nachkauf", { entryId, nachkaufPurchased });
}

export async function removeEntry(entryId: string): Promise<void> {
  await invoke("remove_entry", { entryId });
}

export async function cloneEntries(
  fromCompetitionId: string,
  toCompetitionId: string,
): Promise<CompetitionEntry[]> {
  return invoke("clone_entries", { fromCompetitionId, toCompetitionId });
}

export async function listCompetitionResults(
  competitionId: string,
): Promise<import("@rotpunktarena/domain").EntryResultSummary[]> {
  return invoke("list_competition_results", { competitionId });
}

export async function getEntryResult(
  entryId: string,
): Promise<import("@rotpunktarena/domain").EntryResultDetail | null> {
  return invoke("get_entry_result", { entryId });
}

export async function listEntrySeries(
  entryId: string,
): Promise<import("@rotpunktarena/domain").SeriesResultSummary[]> {
  return invoke("list_entry_series", { entryId });
}

export async function listTeams(
  competitionId?: string | null,
  includeArchived = false,
): Promise<CompetitionTeam[]> {
  return invoke("list_teams", {
    competitionId: competitionId ?? null,
    includeArchived,
  });
}

export async function listKnownTeamNames(
  includeArchived = false,
): Promise<string[]> {
  return invoke("list_known_team_names", { includeArchived });
}

export async function createTeam(
  nameOrCompetitionId: string,
  maybeName?: string,
): Promise<CompetitionTeam> {
  // Back-compat: createTeam(competitionId, name) or createTeam(name)
  const name = maybeName ?? nameOrCompetitionId;
  const competitionId = maybeName != null ? nameOrCompetitionId : null;
  return invoke("create_team", { name, competitionId });
}

export async function renameTeam(teamId: string, name: string): Promise<CompetitionTeam> {
  return invoke("rename_team", { teamId, name });
}

export async function setTeamArchived(
  teamId: string,
  archived: boolean,
): Promise<CompetitionTeam> {
  return invoke("set_team_archived", { teamId, archived });
}

export async function removeTeam(teamId: string): Promise<void> {
  await invoke("remove_team", { teamId });
}

export async function addTeamMember(
  teamId: string,
  entryId: string,
): Promise<CompetitionTeam> {
  return invoke("add_team_member", { teamId, entryId });
}

export async function removeTeamMember(
  teamId: string,
  entryId: string,
): Promise<CompetitionTeam> {
  return invoke("remove_team_member", { teamId, entryId });
}

export async function addTeamPerson(
  teamId: string,
  personId: string,
): Promise<CompetitionTeam> {
  return invoke("add_team_person", { teamId, personId });
}

export async function removeTeamPerson(
  teamId: string,
  personId: string,
): Promise<CompetitionTeam> {
  return invoke("remove_team_person", { teamId, personId });
}

export async function listTeamResults(
  competitionId: string,
): Promise<TeamResultSummary[]> {
  return invoke("list_team_results", { competitionId });
}
